gson/src/main/java/com/google/gson/JsonPrimitive.java

```diff
@@ -290,11 +290,10 @@ public boolean equals(Object obj) {
           : this.getAsNumber().longValue() == other.getAsNumber().longValue();
     }
     if (value instanceof Number && other.value instanceof Number) {
-      double a = getAsNumber().doubleValue();
-      // Java standard types other than double return true for two NaN. So, need
-      // special handling for double.
-      double b = other.getAsNumber().doubleValue();
-      return a == b || (Double.isNaN(a) && Double.isNaN(b));
+      return this.value instanceof BigDecimal || other.value instanceof BigDecimal
+          ? this.getAsBigDecimal().equals(other.getAsBigDecimal())
```

Comment:
One problem might be that `BigDecimal.equals` is quite strict because it also considers the scale of the number (see [documentation](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/math/BigDecimal.html#equals(java.lang.Object))). While this might be fine when both numbers are a `BigDecimal`, maybe this code should be more lenient when one value is not a `BigDecimal` by using `BigDecimal.compareTo`.

For example, the following might be the expected behavior (though maybe other users consider that differently):
```java
assertThat(new JsonPrimitive(new BigDecimal("0.0")).equals(
    new JsonPrimitive(new BigDecimal("0.00")))).isFalse();

String numberString = "0.00";
assertThat(new JsonPrimitive(new BigDecimal(numberString)).equals(
    new JsonPrimitive(Double.valueOf(numberString)))).isTrue();
```
(with the code of this pull request the first assertion passes, but the second one fails)